### DISCUSSION TITLE
Release `rust-bitcoin v0.33.0`

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2753] [Announcements]
~Flagging the intended release date for the next release - end of September 2024.~ I've added a milestone, if you have issues you'd like to see closed before the next release please add them to the milestone.

EDIT: This release is on ice until we have `primitives v1.0.0` released (see https://github.com/rust-bitcoin/rust-bitcoin/discussions/3112 for context).

**Comment by tcharding** [comment]
Heads up, there is close to zero chance that we hit this goal. First we have to release `primitives` and along the way we intend on doing an overhaul of the `hashes` API. (Please accept my humble apologies, some of this was due to me dropping the ball and getting lost in `hashes` for a month or so.)

If you are waiting on any functionality please say so in this thread and if possible we may be able to add and/or back port it.

**Reply by tnull** [reply]
FWIW, at least from my perspective, this is good news as we/downstream projects are struggling anyways to keep up with the recent releases and the amount of breaking API changes they introduced. For projects that depend on multiple dependents of `rust-bitcoin`, it's a constant game of Twister to figure out when we can (and have capacity to) upgrade to what version in order to maintain compatibility with adjacent projects and are able to get to the (other dependent's) features we actually require to make progress. IMO, better coordination of upgrade cycles in the wider `rust-bitcoin` ecosystem would be a great and important topic to discuss in-person at the upcoming meetup.

**Reply by Kixunil** [reply]
@tnull that's why we're pushing to get the most important pieces stabilized so that no coordination will be needed.

**Reply by apoelstra** [reply]
@tnull thanks for the feedback. I think that we should avoid hewing to the deadline for this cycle (within reason -- I don't want to be six+ months late) and focus on getting `primitives` (which is what most downstream projects most care about) into a state where we sincerely think we won't need to break it again. Because as a downstream author the biggest pain point for me is that I'm forced to do huge upgrades all at once due to type incompatibilies between stuff like `Transaction` or `Txid` that haven't even changed in many years. I can't even use two versions of `bitcoin` in parallel because of this.

`hashes` is a bit harder, but as we move stuff out of traits and into generics, that puts us on a better-and-better footing to use the semver trick to avoid "break the world" releases.

Then for everything else we will continue moving more quickly than downstream may be comfortable with ... but it should be far less disruptive.

**Reply by yancyribbens** [reply]
>  we/downstream projects are struggling anyways to keep up with the recent releases and the amount of breaking API changes they introduced.

I'm not sure what there is to keep up with really.  Lot's of the upgrades seem to be purely cosmetic moving crates etc, so i'm not sure how much security critical things there are to "keep up with"?  I think it's only until 1.0 is released that there's really something to keep up with.  I'm sure any feedback between point releases is valuable, so by all means it's good to keep up with the changes in that sense..

**Reply by Kixunil** [reply]
@yancyribbens for instance upgrade to 0.31 was quite significant for Firefish because of much improved APIs. We've deleted a few hundred lines of code. But you're right that there aren't that many urgent things lately and we backport important stuff.

**Comment by tcharding** [comment]
Thanks @tnull please do keep giving us any and all feedback, especially negative. We won't always change our path but its really useful to work out trade offs.

**Comment by tcharding** [comment]
This release is on ice until we have `primitives v1.0.0` released.

**Reply by Kixunil** [reply]
Since this will postpone the release significantly I wonder if we should take a more aggressive approach: make `primitives` just PODs, - only conversion methods, don't wait on `push_decode` - skip having `compute_txid` method or anything similar. Then figure out the script tagging.

But thinking about it I don't think the entire idea of waiting is good at all. We will add new stuff (tagging) to `primitives` and we need the ecosystem to test it. Such is impossible without releasing `primitives` 0.100 and upgrading everything.

**Reply by apoelstra** [reply]
If we release `primitives` without any ability to encode, decode, compute txids, or serialize to hex, then nobody will be able to test them.

In any case, we should be able to get plenty of testing from the rust-elements/simplicity ecosystem, rust-miniscript, our personal projects, etc. But at the rust-bitcoin meetup the maintainers of ldk and bdk were pretty clear that they did not want to do a bunch of extra API-testing and refactoring work on top of their existing workload, when we weren't even close to our final APIs.

**Reply by apoelstra** [reply]
If your point is that we should release rust-bitcoin 0.33 even though we don't expect bdk or ldk to use it, then yeah, I agree with that. My personal wallet, for example, is a good candidate to test `primitives` and `addresses` but it also depends on the BIP32 API which will be stuck in `rust-bitcoin` for a little while.

**Reply by Kixunil** [reply]
> without any ability

It'll still be possible via extension traits, just a bit annoying. Also trivial conversions are fine

> or serialize to hex

I'd definitely make an exception here but hex *deserialization* is a bit hairy because of `Error` type. (Unless we wrap it but then we can't share the type, I think.)

> If your point is that we should release rust-bitcoin 0.33 even though we don't expect bdk or ldk to use it, then yeah, I agree with that.

It wasn't entirely the point but given the circumstances it looks reasonable. For instance Firefish will surely benefit from this since recently I remembered we want `NormalDerivationPath` and we don't depend on BDK.

However I'd still like to look into upgrading LDK and BDK in our forks just to widen the testing. Whether they accept a PR or not is their choice. In best case scenario we find out that everything works as intended and we just bump versions.

**Reply by apoelstra** [reply]
All sounds good. Though let's make sure it actually *is* possible with extension traits or trivial conversions. (I think so, since all our POD types have public fields so we can get to the underlying raw data.)

As for bdk/ldk, agreed that we should fork them and try to upgrade them to `primitives` as we work. But it'll be way too much work to try to keep up with them in real time, so we should just periodically throw away the forks and redo them probably.

**Reply by Kixunil** [reply]
> I think so, since all our POD types have public fields so we can get to the underlying raw data.

Yes, this is the fundamental assumption I made. (`Witness` is a bit more complicated but ultimately it still has obvious interface.)

> As for bdk/ldk, agreed that we should fork them and try to upgrade them to `primitives` as we work. But it'll be way too much work to try to keep up with them in real time, so we should just periodically throw away the forks and redo them probably.

I actually think it won't be that much work to upgrade except for `bip32` changes. That supports your position of course but probably even conflicts won't be that terrible. (Juse `use` statements. For which I have a tool. :D)
